### PR TITLE
LPS-173137: Support PATCH and UPDATE by parentTaxonomyCategory and taxonomyVocabularyId for TaxonomyCategory entity

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
- Add support `PATCH` and `UPDATE` `TaxonomyCategory` entity, allowing you to modify its `parentTaxonomyCategory` and `taxonomyVocabularyId`
- Change `parentTaxonomyCategory` param to not be _readOnly_.

**How to reproduce it?**
1. Deploy `headless-admin-taxonomy-api` and `headless-admin-taxonomy-impl`.
2. Create two Vocabularies: `V1` and `V2`,
3. Create two Categories in `V1`: `Cat1` and `Cat2`.
4. If you want to add `Cat2` as a subcategory of `Cat1`, you have to execute this operation:

```
curl -X 'PATCH' \
  'http://localhost:8080/o/headless-admin-taxonomy/v1.0/taxonomy-categories/{Cat1.Id}' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -u 'test@liferay.com:test
  -d '{
  "parentTaxonomyCategory": {
    "id": ${Cat2.Id}
  }
}'
```
5. If you want to move `Cat1` into of `V2`, you have to execute this operation:
```
curl -X 'PATCH' \
  'http://localhost:8080/o/headless-admin-taxonomy/v1.0/taxonomy-categories/{Cat1.Id}' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -u 'test@liferay.com:test
  -d '{
  "taxonomyCategoryId": ${V2.Id}
 
}'
```
**JIRA Ticket**
https://issues.liferay.com/browse/LPP-47681
https://issues.liferay.com/browse/LPS-173137

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by magjed4289** [comment]
ci:test:relevant

File: modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java [hunk 1]
```diff
@@ -524,6 +526,21 @@ private long _getTotalCount(Long siteId) {
 		return _assetCategoryLocalService.dynamicQueryCount(dynamicQuery);
 	}
 
+	private long _getVocabularyId(
+			AssetCategory assetCategory, TaxonomyCategory taxonomyCategory)
+		throws Exception {
+
+		if (taxonomyCategory.getTaxonomyVocabularyId() != null) {
```
Comment (4lejandrito) [inline modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java]: @sergiojimcos,

Can't we do just:

```java
if (taxonomyCategory.getTaxonomyVocabularyId() != null) {	
	return taxonomyCategory.getTaxonomyVocabularyId();
}

return assetCategory.getVocabularyId();
```

File: modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java [hunk 2]
```diff
@@ -479,6 +484,29 @@ private Page<TaxonomyCategory> _getCategoriesPage(
...
+			AssetCategory assetCategory, TaxonomyCategory taxonomyCategory)
+		throws Exception {
+
+		ParentTaxonomyCategory parentTaxonomyCategory =
+			taxonomyCategory.getParentTaxonomyCategory();
+
+		if ((parentTaxonomyCategory != null) &&
+			(parentTaxonomyCategory.getId() > 0)) {
+
+			AssetCategory newAssetCategory =
```
Comment (4lejandrito) [inline modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java:497]: Shouldn't this be `existingAssetCategory`?

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
Hey @4lejandrito! I couldn't create a test case using `parentTaxonomyCategory`. If you see in the schema, if you want to pass a `parentTaxonomyCategory` you have to create `ParentTaxonomyCategory` object. However, this class is in client module, and the entityResource created doesn't accept other entity objects :man_shrugging:

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java [hunk 3]
```diff
@@ -524,6 +561,23 @@ private long _getTotalCount(Long siteId) {
...
 	}
 
+	private long _getVocabularyId(
+			AssetCategory assetCategory, TaxonomyCategory taxonomyCategory)
+		throws Exception {
+
+		if ((taxonomyCategory.getTaxonomyVocabularyId() != null) &&
+			(taxonomyCategory.getTaxonomyVocabularyId() > 0)) {
+
+			AssetVocabulary newAssetVocabulary =
```
Comment (4lejandrito) [inline modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java:571]: @sergiojimcos this should be renamed to `existing` too.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

**Comment by 4lejandrito** [comment]
> Hey @4lejandrito! I couldn't create a test case using `parentTaxonomyCategory`. If you see in the schema, if you want to pass a `parentTaxonomyCategory` you have to create `ParentTaxonomyCategory` object. However, this class is in `client` module, and the entityResource created doesn't accept other entity objects in a request 🤷‍♂️. I've tried to find any example and there isn't any example that can pass more than one object

I could do it. See https://github.com/liferay-headless/liferay-portal/pull/982/commits/4c8ae7d6223d919cc3bbd7cc8a1101c889835126.

File: modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java (hunk 3 above)
Comment (4lejandrito) [reply modules/apps/headless/headless-admin-taxonomy/headless-admin-taxonomy-impl/src/main/java/com/liferay/headless/admin/taxonomy/internal/resource/v1_0/TaxonomyCategoryResourceImpl.java]: See https://github.com/liferay-headless/liferay-portal/pull/982/commits/8489abe12230eed539b1ee02fbb14e577f1fec12.

**Comment by 4lejandrito** [comment]
ci:forward

**Review by 4lejandrito — APPROVED** [review]

**Pull request closed without merging**